internal/release/observed.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package release
+
+import (
+	"encoding/json"
+	"io"
+
+	"github.com/mitchellh/copystructure"
+	"helm.sh/helm/v3/pkg/chart"
+	helmrelease "helm.sh/helm/v3/pkg/release"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	helmv2 "github.com/fluxcd/helm-controller/api/v2beta2"
+	"github.com/fluxcd/helm-controller/internal/chartutil"
+	"github.com/fluxcd/helm-controller/internal/digest"
+)
+
+var (
+	DefaultDataFilters = []DataFilter{
+		IgnoreHookTestEvents,
+	}
+)
+
+// DataFilter allows for filtering data from the returned ObservedRelease while
+// making an observation.
+type DataFilter func(rel *ObservedRelease)
+
+// IgnoreHookTestEvents ignores test event hooks. This allows to exclude it
+// while e.g. generating a digest for the object. To prevent manual test
+// triggers from a user to interfere with the checksum.
+func IgnoreHookTestEvents(rel *ObservedRelease) {
+	if len(rel.Hooks) > 0 {
+		var hooks []helmrelease.Hook
+		for _, h := range rel.Hooks {
+			if !IsHookForEvent(&h, helmrelease.HookTest) {
+				hooks = append(hooks, h)
+			}
+		}
+		rel.Hooks = hooks
+	}
+}
+
+// ObservedRelease is a copy of a Helm release object, as observed to be written
+// to the storage by a storage.Observer. The object is detached from the Helm
+// storage object, and mutations to it do not change the underlying release
+// object.
+type ObservedRelease struct {
```

Comment:
After going through the code, I haven't seen this object mutated. Will this happen in the reconciler?

Reply:
We do not, it's noted for documentation purposes.